### ISSUE TITLE
git error: "fatal: BUG: disabling cancellation: Invalid argument"

### ISSUE THREAD

**Issue by neclepsio** [Alexpux/MINGW-packages#3351]
Since some updates, when I try to use `git pull` I get the error `fatal: BUG: disabling cancellation: Invalid argument`. I use a proxy.

**Comment by neclepsio** [comment]
Reinstalling with `pacman -S git` seems to have solved.

**Comment by neclepsio** [comment]
Reinstallation seemed to fix, but the problem reappeared and this time reinstallation doesn't fix. The problem is present even if the proxy configuration is deleted.

**Comment by pal1000** [comment]
Tracked upstream as git-for-windows/git#1481.

**Comment by neclepsio** [comment]
The error message is different. Are you sure it's the same problem?

**Comment by pal1000** [comment]
Yes the error messages are related. Differences are only made by presence and absence of proxy and probably Git for Windows running in a slimmer MSYS2 environment.

**Comment by henkman** [comment]
Same problem here with git version 2.16.2. pull, push, log all give the error message
"fatal: BUG: disabling cancellation: Invalid argument".
If I use the msys2.exe shell they work but in cmd.exe the error message occurs

**Comment by pal1000** [comment]
I can repro. I don't know what happens at your end.

**Comment by henkman** [comment]
Just found out that it works in cmd.exe if you have msys2.exe running but not using it

**Comment by sroller** [comment]
It seems the problem was solve upstream. Will there be a new package for msys2 available? I'm running 2.16.2 (64bit) and I'm still having the described problem.

**Comment by ankostis** [comment]
> Just found out that it works in cmd.exe if you have msys2.exe running but not using it. As soon as you close the msys2 shell - boom it does not work anymore.

Same on me with MSYS2, git-2.17.0.  I need to have msys2 or mingw64/32 windows open, or certain git commands (ie `git log`, `diff` ) fail with the above error.  But `git --version` or `git status` `add` run fine.

Very strange.

**Comment by pal1000** [comment]
@sroller, 2.17.0 is available but is not posted [here](https://git-scm.com/download/win) for some reason. A more direct lunk would be: 
[https://github.com/git-for-windows/git/releases/latest](https://github.com/git-for-windows/git/releases/latest).

**Comment by ankostis** [comment]
I note that the problem happens also with older `git version 2.15.0`.

**Comment by axbender** [comment]
Any news on this? This bug keeps nagging me even in git 2.18.0. The same constellation (my MSys2 installation is run from a USB drive) works with Windows 7 while it fails in Windows 10 (1709).

**Comment by rcorvus** [comment]
dmytrol's solution fixed it for me, I removed the msys64 from PATH and after rebooting, git works again.

**Comment by neclepsio** [comment]
dmytrol's solution DID NOT fix for me, as the path was right.

In my case as henkman writes it works in cmd.exe if msys window is open (https://github.com/Alexpux/MINGW-packages/issues/3351#issuecomment-368065275).

**Comment by AnujaySaraf** [comment]
Hi All,

Downgrading the version of git to 2.16.1 helps this issue, one can downgrade the Git version by downloading and installing the same in Windows.

link;- https://github.com/git-for-windows/git/releases/tag/v2.16.1.windows.1

Regards,
Anujay

**Comment by wangxuesong29** [comment]
**Solved!**

 If you recently deal with my `MSYS2`:

**delete the environment path of it.**

**Comment by axbender** [comment]
@wangxuesong29
Could you please detail on what you mean exactly?

**Comment by wangxuesong29** [comment]
@axbender 
Yesterday I added `C:\msys64\msys2.exe` in `PATH` of  `Environment variable`.  Then the error occurs.

I deleted it from the  `PATH`. 
![image](https://user-images.githubusercontent.com/14146871/60768057-db9a5380-a0c0-11e9-8815-13993d42c901.png)

it might be a coincidence, if you dealt with `msys` recently you can check it. Because many people mentioned `msys` before, I got an idea from it.

**Comment by axbender** [comment]
@wangxuesong29

Hi!
Unfortunately this doesn't work on my end... Also, please consider:
a) Git works w/o problems on Windows 7 using the exact same configuration (USB drive).
b) Adding an EXE to your PATH environment variable is useless (it's no path); the system will not honor it (lest your "msys2.exe" is a path... ;-).

Once I read something about Windows 10 having problems with the relocation addresses of MSys2 executables. I followed the recipe, but it wouldn't work either. I'm still at loss...

**Comment by wangxuesong29** [comment]
@axbender 
Hi.

I tried to use msys2 again, so I add `C:\msys64\usr\bin` into environment variables, which caused the error again.

I then realize that there are two `git`s, one from git you installed before, one from `msys2`(you might ran `pacman -S git patch unzip` before you met this error.)

Solution:

I uninstall the `git` installed by me, which should be called like `git-for-windows`. Because I need to used `msys2`.

**Comment by ankostis** [comment]
Still we have not isolated the root-cause of the issue, and why one installation affects the other.

**Comment by axbender** [comment]
No double installation here, just MSys2's git...

**Comment by neclepsio** [comment]
Just msys2 installation for me, and it doesn't work from any shell if msys2 shell is not open. But if I open a msys2 shell, it works even from every shell.

**Comment by gbrunick** [comment]
I see the same behavior as @neclepsio. `git fetch` works if I have a Mintty Bash window open (MSYS2 or Ming2-w64 64bit), and fails with 
```
> git fetch
BUG: run-command.c:519: disabling cancellation: Invalid argument
```
otherwise.  I'm on Windows 10, I use a proxy as well, and:
```
> git --version
git version 2.24.0
> which git
/usr/bin/git
```

**Comment by 4sfaloth** [comment]
I have no clue about the underlying causes, but if it helps someone I was facing this issue (**_fatal: BUG disabling cancellation: Invalid argument_** when trying to run **_git branch --merged_**) and it was fixed by following @wangxuesong29's suggestion (ie: removing c:\msys64\usr\bin from path). Not sure what the consequences of this for other applications might be though...

**Comment by sroller** [comment]
Are you sure there is no other git.exe in your path?

**Comment by 4sfaloth** [comment]
> Are you sure there is no other git.exe in your path?

I cannot guarantee that. I can only say that I was experiencing the fatal: BUG described above when trying to run git branch --merged and that it was fixed by removing **_c:\msys64\usr\bin_** from the path as per @wangxuesong29 suggestion

**Comment by axbender** [comment]
No "double" git.exe here; however, in my MSys2 installation there's a hard link between /usr/bin/git.exe and /bin/git.exe, and there's a /mingw64/bin/git which calls /usr/bin/git. I renamed/copied/... several combinations but the error remains (other than in Windows 7)

**Comment by StarWolf3000** [comment]
> No "double" git.exe here; however, in my MSys2 installation there's a hard link between /usr/bin/git.exe and /bin/git.exe, and there's a /mingw64/bin/git which calls /usr/bin/git. I renamed/copied/... several combinations but the error remains (other than in Windows 7)

There's no `/bin` directory in MSYS2.

**Comment by axbender** [comment]
Hm, I'm tempted to say "you're wrong" (as I do have a /bin and a /usr/bin directory), but may be this is the core of the problem. Here both directories are hardlinked. How would I get rid of this association? Just rm -rf /bin seems to be too dangerous...

**Comment by axbender** [comment]
Addendum: /bin does not exist outside of MSys2 (e.g. using Windows Explorer), but is visible insided mintty.

**Comment by StarWolf3000** [comment]
According to the `mount` command, `/bin` is mounted from `/usr/bin` so doing the `rm -rf /bin` would kill your MSYS2 environment, except for `bash.exe`, `mintty.exe`, `rm.exe` and several dlls.

**Comment by axbender** [comment]
Interesting. Who would mount /bin to /usr/bin? Checked /etc/fstab => nope.
Any idea?

**Comment by axbender** [comment]
Update: Found this in github/MSys2:

Alexx83 - 2014-06-24

Now we remove mounting /usr to / in MSYS2.
Today we switch to new MSYS2 files layout. Now all MSYS2 packages installing into real /usr prefix.
Root directory contains only sysconfig directories: etc, var, tmp, home. 
For mingw-w64 packages nothing changed - they installing as usual in to /mingw32 and /mingw64 folders.
**But we still have one virtual mount directory - for compatibility reasons we mount /bin to /usr/bin.**

**Comment by StarWolf3000** [comment]
I think that's still done to keep high compatibiliy with Cygwin, which MSYS2 is based on.

**Comment by axbender** [comment]
Any news on the original topic?

**Comment by Gei0r** [comment]
The problem still exists when running `git pull` from the windows `cmd` shell.

```
git version 2.26.2
BUG: run-command.c:519: disabling cancellation: Invalid argument
```

Curiously, if a msys2 shell (`C:\msys64\usr\bin\bash.exe` in my case) is running, `git pull` also works from a `cmd` shell.

**Comment by StarWolf3000** [comment]
Still cannot reproduce any problems with Git. I can use it just fine from both CMD and MSYS2 and MINGW shells.

**Comment by Gei0r** [comment]
Is there a way I can produce more debugging information on this problem?

**Comment by beliaev-maksim** [comment]
hey guys,
once you install MSYS2 you automatically add env variable that points to bin of MSYS. After upgrade of git you may get your git installation in PATH env var be later in the list.

Once git is called in terminal, shell looks for executable in path and finds MSYS2 git first.
Just edit your environment variables the way, that first comes "git" and then "MSYS2"
![image](https://user-images.githubusercontent.com/51964909/113690281-85aa7380-96cb-11eb-8cbb-89232aaec03d.png)

**Comment by duocang** [comment]
I have deleted my old answer.

### Simple answer: There are two `git.exe`, they fight.
- You installed one
- one comes with msys.

### What to do?
@beliaev-maksim 
Once git is called in terminal, shell looks for executable in path and finds MSYS2 git first.
Just edit your environment variables the way, that first comes "git" and then "MSYS2"
![](https://user-images.githubusercontent.com/51964909/113690281-85aa7380-96cb-11eb-8cbb-89232aaec03d.png)

**Comment by axbender** [comment]
Probably not. 
I‘m using the identical portable installation from my USB drive on my Windows 7 (ok), my Windows 10 (ok), and my company Windows 10 (fails).
I guess - meanwhile - that the company one is heavily locked down (although I am a local admin) in terms of so-called „security features“. I‘ve already tried a lot but to no avail...

**Comment by kimim** [comment]
This error happens also when I invoke git in emacs eshell:

```
Welcome to the Emacs shell

~/kimim-emacs $ git pull
BUG: run-command.c:520: disabling cancellation: Invalid argument
~/kimim-emacs $ which git
c:/msys64/usr/bin/git.exe
~/kimim-emacs $ git version
git version 2.31.1
```

**Comment by rwreay** [comment]
Same issue: Win10 Windows Command Prompt, **git pull** and **git clone** fail in command prompt unless an instance of an MSYS shell is running. There is only one instance of git on this machine.  

Possibly relevant, Strawberry Perl and Miniconda are installed and in the path.

```
[...]>git pull
BUG: run-command.c:520: disabling cancellation: Invalid argument

[...]>where git
C:\msys64\usr\bin\git.exe

[...]>git version
git version 2.32.0
```

**Comment by beliaev-maksim** [comment]
@rwreay 
I am not sure that you can use MSYS git in Win Shell. You have to use MSYS git in MSYS console.

otherwise install git for windows in parallel and ensure proper paths in env vars (see my previous message)

**Comment by Biswa96** [comment]
> I am not sure that you can use MSYS git in Win Shell. 

Yes, you can. MSYS2 programs are like other Win32 programs, nothing special. But you have to make sure that the `%PATH%` is not mixed up with other toolchains, as rwreay mentioned strawberry perl etc.

**Comment by kimim** [comment]
> Same issue: Win10 Windows Command Prompt, **git pull** and **git clone** fail in command prompt unless an instance of an MSYS shell is running. There is only one instance of git on this machine.
> 
Thank you for providing a workaround. I use git in Emacs/magit. When a MingW64 window is open, git works fine also.
I used to think it is a random problem, that git works fine sometimes and gets error randomly..

**Comment by rwreay** [comment]
> But you have to make sure that the %PATH% is not mixed up with other toolchains, as rwreay mentioned strawberry perl etc.

This problem occurs when there is only msys2 directories in my path.

```
[...]>echo %PATH%
C:\msys64\mingw32\bin;C:\msys64\usr\local\bin;C:\msys64\usr\bin

[...]>git pull
BUG: run-command.c:520: disabling cancellation: Invalid argument

[...]>echo starting MSYS shell...
starting MSYS shell...

[...]>git pull
Already up to date.
```

**Comment by ryanavella** [comment]
I'm having the same issue. Strangely, some of the git subcommands work if I redirect the output to a temporary file. For example:

```cmd
C:\tmp\repo>git log
BUG: run-command.c:519: disabling cancellation: Invalid argument

C:\tmp\repo>git log > test && type test
commit a78d08b9bdb401a5ddfb5bea97240af7c49e9795
Merge: 482afb59 fd82aa28
...
```

**Comment by sansalinas** [comment]
In my case, just open msys2_shell.cmd previosly and it works :S

**Comment by malkia** [comment]
I was also looking for solution, and what @sansalinas ^^^ proposed worked for me too, though I wanted to avoid the window being visible, and launched mintty (which seems to launch bash, and probably that made it working) - e.g. in C:\ProgramData\Microsoft\Windows\Start Menu\Programs\Startup (or per user in C:\Users\[User Name]\AppData\Roaming\Microsoft\Windows\Start Menu\Programs\Startup) I've added a batch file that does this: ```
c:\msys64\usr\bin\mintty.exe -w hide```

**Comment by jonas-fsr** [comment]
Hey all,
so in my case the error appeared after I upgraded my Intellij IDE to the latest version (2022.1.3) which appeared not to match the latest version of git and git toolbox plugin. The error vanished after downgrading the IDE back to version 2021.3.3.

**Comment by ryanavella** [comment]
I moved `C:\msys64\usr\bin` to the very top of my PATH, but that didn't quite fix it.

However, the symptoms disappeared after updating to git-2.42.0.
